java/ql/lib/semmle/code/java/dataflow/TypeFlow.qll

```diff
@@ -413,30 +429,41 @@ private predicate downcastSuccessor(VarAccess va, RefType t) {
   )
 }
 
+private Expr getAProbableAlias(Expr e) {
+  exists(BaseSsaVariable v |
+    e = v.getAUse() and
+    result = v.getAUse()
+  )
+  or
+  exists(BaseSsaVariable v1, BaseSsaVariable v2, ArrayAccess aa1, ArrayAccess aa2 |
+    e = aa1 and
+    result = aa2 and
+    aa1.getArray() = v1.getAUse() and
+    aa1.getIndexExpr() = v2.getAUse() and
+    aa2.getArray() = v1.getAUse() and
+    aa2.getIndexExpr() = v2.getAUse()
+  )
+}
+
 /**
- * Holds if `va` is an access to a value that is guarded by `instanceof t`.
+ * Holds if `e` is an access to a value that is guarded by `instanceof t`.
  */
-private predicate instanceOfGuarded(VarAccess va, RefType t) {
-  exists(InstanceOfExpr ioe, BaseSsaVariable v |
-    ioe.getExpr() = v.getAUse() and
+private predicate instanceOfGuarded(Expr e, RefType t) {
+  exists(InstanceOfExpr ioe |
     t = ioe.getCheckedType() and
-    va = v.getAUse() and
-    guardControls_v1(ioe, va.getBasicBlock(), true)
+    e = getAProbableAlias(ioe.getExpr()) and
+    guardControls_v1(ioe, e.getBasicBlock(), true)
   )
 }
 
 /**
- * Holds if `aa` is an access to a value that is guarded by `instanceof t`.
+ * Holds if `e` is an access to a value that is guarded by `case T t`.
  */
-predicate arrayInstanceOfGuarded(ArrayAccess aa, RefType t) {
-  exists(InstanceOfExpr ioe, BaseSsaVariable v1, BaseSsaVariable v2, ArrayAccess aa1 |
-    ioe.getExpr() = aa1 and
-    t = ioe.getCheckedType() and
-    aa1.getArray() = v1.getAUse() and
-    aa1.getIndexExpr() = v2.getAUse() and
-    aa.getArray() = v1.getAUse() and
-    aa.getIndexExpr() = v2.getAUse() and
-    guardControls_v1(ioe, aa.getBasicBlock(), true)
+private predicate patternCaseGuarded(Expr e, RefType t) {
+  exists(PatternCase pc |
+    e = getAProbableAlias([pc.getSwitch().getExpr(), pc.getSwitchExpr().getExpr()]) and
```

Comment:
The pattern `[pc.getSwitch().getExpr(), pc.getSwitchExpr().getExpr()]` already has an existing convenience method `getSelectorExpr`.